File: packages/shared/agent-sync.ts
/**
 * Agent Sync wire types. These mirror the serde output of the Rust crate in
 * packages/agent-sync (camelCase field names, enum variants as camelCase
 * strings). Keep the two in lockstep.
 *
 * CDXC:AgentSync 2026-09-16 SEE-ALSO:
 * packages/agent-sync/src/{scan,plan,apply}.rs are the source of truth for these shapes; the
 * desktop bridge (apps/desktop/src/app/helpers/agents_hub/sync.rs) and `ghostex agent-sync`
 * both emit them unchanged.
 */

export type AgentSyncAgentStatus = 'linked' | 'attention' | 'notInstalled';
export type AgentSyncAgentKind = 'agent' | 'profile';
export type AgentSyncSkillDirState = 'missing' | 'realDir' | 'wholeFolderLink' | 'otherLink' | 'notADir';
export type AgentSyncSkillEntryState =
  | 'linked'
  | 'linkedElsewhere'
  | 'dangling'
  | 'copyIdentical'
  | 'copyDrifted'
  | 'onlyHere'
  | 'missing'
  | 'viaWholeFolder';
export type AgentSyncInstructionKind = 'plain' | 'mdc';
export type AgentSyncInstructionState =
  'missing' | 'pointer' | 'legacyPointer' | 'otherContent' | 'symlink' | 'notAFile';
export type AgentSyncHooksState = 'linked' | 'missing' | 'realDir' | 'otherLink';
export type AgentSyncLockLinkState = 'linked' | 'missing' | 'realFile' | 'otherLink';
export type AgentSyncSourceGitState = 'none' | 'noCommits' | 'committed';

export type AgentSyncSkillEntry = {
  ghostexBundled: boolean;
  linkTarget?: string;
  name: string;
  path: string;
  state: AgentSyncSkillEntryState;
};

export type AgentSyncSkillCounts = {
  copiesDrifted: number;
  copiesIdentical: number;
  dangling: number;
  linked: number;
  linkedElsewhere: number;
  missing: number;
  onlyHere: number;
  viaWholeFolder: number;
};

export type AgentSyncSkillsReport = {
  counts: AgentSyncSkillCounts;
  dir: string;
  dirLinkTarget?: string;
  dirState: AgentSyncSkillDirState;
  entries: AgentSyncSkillEntry[];
};

export type AgentSyncInstructionReport = {
  kind: AgentSyncInstructionKind;
  path: string;
  state: AgentSyncInstructionState;
};

export type AgentSyncHooksReport = {
  dir: string;
  linkTarget?: string;
  state: AgentSyncHooksState;
};

export type AgentSyncLockLinkReport = {
  path: string;
  state: AgentSyncLockLinkState;
};

export type AgentSyncAgentReport = {
  detected: boolean;
  displayName: string;
  hooks?: AgentSyncHooksReport;
  icon?: string;
  id: string;
  instructions?: AgentSyncInstructionReport;
  kind: AgentSyncAgentKind;
  lock?: AgentSyncLockLinkReport;
  root: string;
  skills?: AgentSyncSkillsReport;
  status: AgentSyncAgentStatus;
  universal: boolean;
};

export type AgentSyncSourceSkill = {
  broken: boolean;
  ghostexBundled: boolean;
  inLock: boolean;
  isSymlink: boolean;
  name: string;
  path: string;
};

export type AgentSyncLockInfo = {
  entryCount: number;
  exists: boolean;
  path: string;
  staleEntries: string[];
  untrackedSkills: string[];
};

export type AgentSyncSourceInfo = {
  exists: boolean;
  git: AgentSyncSourceGitState;
  hookScriptCount: number;
  hooksDir: string;
  hooksDirExists: boolean;
  lock: AgentSyncLockInfo;
  mainMdExists: boolean;
  mdFiles: string[];
  path: string;
  skills: AgentSyncSourceSkill[];
  skillsDir: string;
};

export type AgentSyncProblemKind =
  | 'danglingLinks'
  | 'sourceBrokenLinks'
  | 'staleLockEntries'
  | 'untrackedSkills'
  | 'copiedFolders'
  | 'wholeFolderLinks'
  | 'missingPointers';

export type AgentSyncProblem = {
  count: number;
  detail: string;
  fixable: boolean;
  items: string[];
  kind: AgentSyncProblemKind;
  title: string;
};

export type AgentSyncReportSummary = {
  agentsAttention: number;
  agentsDetected: number;
  agentsLinked: number;
  agentsNotInstalled: number;
  copiedSkillFolders: number;
  danglingLinks: number;
  missingPointers: number;
  staleLockEntries: number;
  wholeFolderLinks: number;
};

export type AgentSyncReport = {
  agents: AgentSyncAgentReport[];
  generatedAt: string;
  home: string;
  problems: AgentSyncProblem[];
  source: AgentSyncSourceInfo;
  summary: AgentSyncReportSummary;
};

export type AgentSyncPlanVerb = 'link' | 'write' | 'backup' | 'unlink' | 'keep' | 'mkdir' | 'drop';
export type AgentSyncPlanGroupKind =
  'removeDangling' | 'convertWholeFolder' | 'perSkillLinks' | 'pointerFiles' | 'hooks' | 'pruneLock';

export const AGENT_SYNC_PLAN_GROUP_KINDS: readonly AgentSyncPlanGroupKind[] = [
  'removeDangling',
  'convertWholeFolder',
  'perSkillLinks',
  'pointerFiles',
  'hooks',
  'pruneLock',
];

export type AgentSyncPlanOp = {
  agentId?: string;
  content?: string;
  note?: string;
  path: string;
  target?: string;
  verb: AgentSyncPlanVerb;
};

export type AgentSyncPlanGroup = {
  changeCount: number;
  enabledByDefault: boolean;
  kind: AgentSyncPlanGroupKind;
  ops: AgentSyncPlanOp[];
  title: string;
};

export type AgentSyncPlanSummary = {
  backups: number;
  defaultChanges: number;
  drops: number;
  keeps: number;
  links: number;
  mkdirs: number;
  unlinks: number;
  writes: number;
};

export type AgentSyncPlan = {
  generatedAt: string;
  groups: AgentSyncPlanGroup[];
  scope: string;
  stamp: string;
  summary: AgentSyncPlanSummary;
};

export type AgentSyncApplyFailure = {
  error: string;
  op: AgentSyncPlanOp;
};

export type AgentSyncApplyResult = {
  done: AgentSyncPlanOp[];
  enabledGroups: AgentSyncPlanGroupKind[];
  failed: AgentSyncApplyFailure[];
  generatedAt: string;
  plan: AgentSyncPlan;
  scope: string;
  skippedKeeps: number;
};

/** The plan groups a problem row's fix button enables. */
export function agentSyncProblemFixGroups(kind: AgentSyncProblemKind): AgentSyncPlanGroupKind[] {
  switch (kind) {
    case 'danglingLinks':
    case 'sourceBrokenLinks':
      return ['removeDangling'];
    case 'wholeFolderLinks':
      return ['convertWholeFolder'];
    case 'copiedFolders':
      return ['perSkillLinks'];
    case 'missingPointers':
      return ['pointerFiles'];
    case 'staleLockEntries':
      return ['pruneLock'];
    case 'untrackedSkills':
      return [];
  }
}

export function agentSyncDefaultPlanGroups(plan: AgentSyncPlan | undefined): AgentSyncPlanGroupKind[] {
  if (!plan) {
    return AGENT_SYNC_PLAN_GROUP_KINDS.filter((kind) => kind !== 'pruneLock');
  }
  return plan.groups.filter((group) => group.enabledByDefault).map((group) => group.kind);
}

export function agentSyncPlanChangeCount(plan: AgentSyncPlan, enabled: ReadonlySet<AgentSyncPlanGroupKind>): number {
  return plan.groups.reduce((sum, group) => (enabled.has(group.kind) ? sum + group.changeCount : sum), 0);
}

/** Skills, Instructions, Hook scripts: the three things Agent Sync shares, one tone each. */
export type AgentSyncDotTone = 'ok' | 'warn' | 'err' | 'off';

export function agentSyncSkillsTone(agent: AgentSyncAgentReport): AgentSyncDotTone {
  const skills = agent.skills;
  if (!skills) {
    return agent.universal ? 'ok' : 'off';
  }
  if (skills.counts.dangling > 0 || skills.dirState === 'otherLink' || skills.dirState === 'notADir') {
    return 'err';
  }
  if (
    skills.dirState === 'wholeFolderLink' ||
    skills.counts.copiesIdentical > 0 ||
    skills.counts.copiesDrifted > 0 ||
    (!agent.universal && skills.counts.missing > 0)
  ) {
    return 'warn';
  }
  return 'ok';
}

export function agentSyncInstructionsTone(agent: AgentSyncAgentReport): AgentSyncDotTone {
  const instructions = agent.instructions;
  if (!instructions) {
    return 'off';
  }
  switch (instructions.state) {
    case 'pointer':
      return 'ok';
    case 'missing':
    case 'legacyPointer':
    case 'otherContent':
      return 'warn';
    default:
      return 'err';
  }
}

export function agentSyncHooksTone(agent: AgentSyncAgentReport): AgentSyncDotTone {
  const hooks = agent.hooks;
  if (!hooks) {
    return 'off';
  }
  switch (hooks.state) {
    case 'linked':
      return 'ok';
    case 'missing':
    case 'realDir':
      return 'warn';
    default:
      return 'err';
  }
}

export type AgentSyncPart = 'skills' | 'instructions' | 'hooks';

/** The plan groups that fix one part of one agent; empty when that part has nothing to fix. */
export function agentSyncPartFixGroups(agent: AgentSyncAgentReport, part: AgentSyncPart): AgentSyncPlanGroupKind[] {
  switch (part) {
    case 'skills': {
      const skills = agent.skills;
      if (!skills) {
        return [];
      }
      const groups: AgentSyncPlanGroupKind[] = [];
      if (skills.counts.dangling > 0) {
        groups.push('removeDangling');
      }
      if (skills.dirState === 'wholeFolderLink') {
        groups.push('convertWholeFolder');
      }
      if (agent.detected && (skills.counts.copiesIdentical > 0 || (!agent.universal && skills.counts.missing > 0))) {
        groups.push('perSkillLinks');
      }
      return groups;
    }
    case 'instructions': {
      const state = agent.instructions?.state;
      return agent.detected && (state === 'missing' || state === 'legacyPointer' || state === 'otherContent')
        ? ['pointerFiles']
        : [];
    }
    case 'hooks': {
      const hooksOpen = agent.hooks && agent.hooks.state !== 'linked' && agent.hooks.state !== 'otherLink';
      const lockOpen = agent.lock && agent.lock.state !== 'linked' && agent.lock.state !== 'otherLink';
      return agent.detected && (hooksOpen || lockOpen) ? ['hooks'] : [];
    }
  }
}

export const AGENT_SYNC_PARTS: readonly AgentSyncPart[] = ['skills', 'instructions', 'hooks'];

/** How many fixes an agent row advertises ("3 to fix"). An agent the scan flags always shows at least one. */
export function agentSyncIssueCount(agent: AgentSyncAgentReport): number {
  const count = AGENT_SYNC_PARTS.reduce((sum, part) => sum + agentSyncPartFixGroups(agent, part).length, 0);
  return agent.status === 'attention' ? Math.max(count, 1) : count;
}

/** Red when something is broken, amber when it works today but can go out of date. */
export function agentSyncWorstTone(agent: AgentSyncAgentReport): 'ok' | 'warn' | 'err' {
  const tones = [agentSyncSkillsTone(agent), agentSyncInstructionsTone(agent), agentSyncHooksTone(agent)];
  if (tones.includes('err')) {
    return 'err';
  }
  return tones.includes('warn') || agent.status === 'attention' ? 'warn' : 'ok';
}

/** Which of the three shared things a problem belongs to; the shared folder's own problems have none. */
export function agentSyncProblemPart(kind: AgentSyncProblemKind): AgentSyncPart | undefined {
  switch (kind) {
    case 'danglingLinks':
    case 'copiedFolders':
    case 'wholeFolderLinks':
      return 'skills';
    case 'missingPointers':
      return 'instructions';
    default:
      return undefined;
  }
}

export const AGENT_SYNC_SKILL_ENTRY_LABELS: Record<AgentSyncSkillEntryState, string> = {
  copyDrifted: 'Copy, differs from the source',
  copyIdentical: 'Copy, same content as the source',
  dangling: 'Broken link',
  linked: 'Linked',
  linkedElsewhere: 'Linked to another folder',
  missing: 'Not linked yet',
  onlyHere: 'Only here',
  viaWholeFolder: 'Available through the folder link',
};

export const AGENT_SYNC_INSTRUCTION_LABELS: Record<AgentSyncInstructionState, string> = {
  legacyPointer: 'Older pointer wording',
  missing: 'Not present',
  notAFile: 'Not a file',
  otherContent: 'Has content of its own',
  pointer: 'Points at ~/.agents/main.md',
  symlink: 'Is a symlink',
};

export const AGENT_SYNC_VERB_LABELS: Record<AgentSyncPlanVerb, string> = {
  backup: 'backup',
  drop: 'drop',
  keep: 'keep',
  link: 'link',
  mkdir: 'mkdir',
  unlink: 'unlink',
  write: 'write',
};

File: server/src/ghostex_cli/agent_sync.rs
use serde_json::{json, Value};

use crate::ghostex_cli::args::parse_args;
use crate::ghostex_cli::output::print_json;
use crate::ghostex_cli::rpc::{CliError, CliResult};
use crate::paths::get_gxserver_paths;
use ghostex_agent_sync::{
    apply, build_plan, scan, AgentStatus, PlanGroupKind, PlanOptions, PlanVerb, SyncPlan,
    SyncReport, SyncScope,
};

/*
CDXC:AgentSync 2026-09-16 DECISION:
User: the CLI verb ships together with the Agents Hub tab and both share one crate (3A),
so an agent, a remote host, or a script gets exactly the scan and plan the Hub shows.
`apply` refuses to run without --yes because it renames folders and writes files.
*/

pub fn agent_sync_usage() -> String {
    "Ghostex Agent Sync - point every agent at the shared ~/.agents folder

Usage:
  gx agent-sync status [--json]
  gx agent-sync plan [--agent <id>] [--group <group>...] [--prune-lock] [--json]
  gx agent-sync apply --yes [--agent <id>] [--group <group>...] [--prune-lock] [--json]
  gx agent-sync agents [--json]

What it does:
  ~/.agents is the source of truth: skills/, main.md and the other rule files,
  hooks/, and .skill-lock.json. Sync creates one relative symlink per skill in
  every agent's global skills folder, converts whole-folder links into folders
  of per-skill links, writes the one-line instruction pointer file each agent
  reads, and links the hooks folder and lock file into Claude Code and Codex.
  Nothing is deleted: anything in the way of a link is renamed to
  <name>.pre-sync-<stamp>.bak, and dangling links are removed.

Groups (all on by default except prune-lock):
  remove-dangling        Remove symlinks whose target is gone
  convert-whole-folder   Replace a skills folder symlink with per-skill links
  per-skill-links        Link every source skill into the agent folder
  pointer-files          Write the instruction pointer files
  hooks                  Link the hooks folder and .skill-lock.json
  prune-lock             Drop lock entries that have no skill folder (opt-in)

Options:
  --agent <id>           One agent (see `gx agent-sync agents`), default all
  --group <group>        Run only these groups (repeatable); default: every group
                         that is on by default
  --prune-lock           Also run the prune-lock group
  --yes                  Required by apply
  --json                 Print JSON
"
    .to_string()
}

fn wants_help(args: &[String]) -> bool {
    args.is_empty() || matches!(args[0].as_str(), "help" | "-h" | "--help")
}

fn home_dir() -> CliResult<std::path::PathBuf> {
    let paths = get_gxserver_paths(None);
    let home = paths.agent_config_home_dir().to_path_buf();
    if home.as_os_str().is_empty() {
        return Err(CliError::Other(
            "Could not determine the home folder.".to_string(),
        ));
    }
    Ok(home)
}

fn parse_groups(args: &[String], flags_prune: bool) -> CliResult<Vec<PlanGroupKind>> {
    let mut groups: Vec<PlanGroupKind> = Vec::new();
    let mut index = 0;
    while index < args.len() {
        if args[index] == "--group" {
            if let Some(value) = args.get(index + 1) {
                let kind = PlanGroupKind::parse(value)
                    .ok_or_else(|| CliError::Other(format!("Unknown agent-sync group: {value}")))?;
                if !groups.contains(&kind) {
                    groups.push(kind);
                }
                index += 2;
                continue;
            }
        }
        if let Some(value) = args[index].strip_prefix("--group=") {
            let kind = PlanGroupKind::parse(value)
                .ok_or_else(|| CliError::Other(format!("Unknown agent-sync group: {value}")))?;
            if !groups.contains(&kind) {
                groups.push(kind);
            }
        }
        index += 1;
    }
    if groups.is_empty() {
        groups = PlanGroupKind::ALL
            .iter()
            .copied()
            .filter(|kind| *kind != PlanGroupKind::PruneLock)
            .collect();
    }
    if flags_prune && !groups.contains(&PlanGroupKind::PruneLock) {
        groups.push(PlanGroupKind::PruneLock);
    }
    Ok(groups)
}

pub fn agent_sync_command(args: &[String]) -> CliResult<()> {
    if wants_help(args) {
        println!("{}", agent_sync_usage());
        return Ok(());
    }
    let verb = args[0].as_str();
    let rest: Vec<String> = args[1..].to_vec();
    let parsed = parse_args(&rest);
    let json = parsed.flags.truthy("json");
    let home = home_dir()?;
    let scope = SyncScope::parse(parsed.flags.text("agent").as_deref());
    match verb {
        "status" => {
            let report = scan(&home);
            if json {
                print_json(
                    &serde_json::to_value(&report)
                        .map_err(|error| CliError::Other(error.to_string()))?,
                );
            } else {
                print_status(&report);
            }
            Ok(())
        }
        "agents" => {
            let report = scan(&home);
            if json {
                let agents: Vec<Value> = report
                    .agents
                    .iter()
                    .map(|agent| {
                        json!({
                            "id": agent.id,
                            "displayName": agent.display_name,
                            "detected": agent.detected,
                            "status": agent.status,
                            "root": agent.root,
                        })
                    })
                    .collect();
                print_json(&Value::Array(agents));
            } else {
                for agent in report.agents.iter().filter(|agent| agent.detected) {
                    println!(
                        "{:<28} {:<10} {}",
                        agent.id,
                        status_label(agent.status),
                        agent.root
                    );
                }
            }
            Ok(())
        }
        "plan" => {
            let report = scan(&home);
            let options = PlanOptions {
                scope,
                ..PlanOptions::default()
            };
            let plan = build_plan(&home, &report, &options);
            let groups = parse_groups(&rest, parsed.flags.truthy("pruneLock"))?;
            if json {
                print_json(
                    &serde_json::to_value(&plan)
                        .map_err(|error| CliError::Other(error.to_string()))?,
                );
            } else {
                print_plan(&plan, &groups);
            }
            Ok(())
        }
        "apply" => {
            let groups = parse_groups(&rest, parsed.flags.truthy("pruneLock"))?;
            if !parsed.flags.truthy("yes") {
                let report = scan(&home);
                let options = PlanOptions {
                    scope,
                    ..PlanOptions::default()
                };
                let plan = build_plan(&home, &report, &options);
                print_plan(&plan, &groups);
                return Err(CliError::Other(
                    "Nothing was changed. Re-run with --yes to apply this plan.".to_string(),
                ));
            }
            let options = PlanOptions {
                scope,
                ..PlanOptions::default()
            };
            let result = apply(&home, &options, &groups);
            if json {
                print_json(
                    &serde_json::to_value(&result)
                        .map_err(|error| CliError::Other(error.to_string()))?,
                );
            } else {
                println!(
                    "Applied {} operation(s), {} failed, {} already correct.",
                    result.done.len(),
                    result.failed.len(),
                    result.skipped_keeps
                );
                for failure in &result.failed {
                    println!(
                        "  FAILED {:<7} {} ({})",
                        verb_label(failure.op.verb),
                        failure.op.path,
                        failure.error
                    );
                }
            }
            if result.failed.is_empty() {
                Ok(())
            } else {
                Err(CliError::Other(format!(
                    "{} operation(s) failed.",
                    result.failed.len()
                )))
            }
        }
        other => Err(CliError::Other(format!(
            "Unknown agent-sync command: {other}\n\n{}",
            agent_sync_usage()
        ))),
    }
}

fn status_label(status: AgentStatus) -> &'static str {
    match status {
        AgentStatus::Linked => "linked",
        AgentStatus::Attention => "attention",
        AgentStatus::NotInstalled => "missing",
    }
}

fn verb_label(verb: PlanVerb) -> &'static str {
    match verb {
        PlanVerb::Link => "link",
        PlanVerb::Write => "write",
        PlanVerb::Backup => "backup",
        PlanVerb::Unlink => "unlink",
        PlanVerb::Keep => "keep",
        PlanVerb::Mkdir => "mkdir",
        PlanVerb::Drop => "drop",
    }
}

fn print_status(report: &SyncReport) {
    let source = &report.source;
    println!(
        "Source {}: {} skills, {} md files, {} hook scripts, lock {} entries, git {:?}",
        source.path,
        source.skills.len(),
        source.md_files.len(),
        source.hook_script_count,
        source.lock.entry_count,
        source.git
    );
    let summary = &report.summary;
    println!(
        "Agents: {} detected, {} linked, {} need attention. Dangling links {}, copied folders {}, whole-folder links {}, missing pointers {}, stale lock entries {}.",
        summary.agents_detected,
        summary.agents_linked,
        summary.agents_attention,
        summary.dangling_links,
        summary.copied_skill_folders,
        summary.whole_folder_links,
        summary.missing_pointers,
        summary.stale_lock_entries
    );
    println!();
    for agent in report.agents.iter().filter(|agent| agent.detected) {
        let skills = agent
            .skills
            .as_ref()
            .map(|skills| {
                let c = &skills.counts;
                format!(
                    "skills {:?} linked {} dangling {} copies {} only-here {} missing {}",
                    skills.dir_state,
                    c.linked + c.via_whole_folder,
                    c.dangling,
                    c.copies_identical + c.copies_drifted,
                    c.only_here,
                    c.missing
                )
            })
            .unwrap_or_else(|| "skills n/a".to_string());
        let instructions = agent
            .instructions
            .as_ref()
            .map(|item| format!("md {:?}", item.state))
            .unwrap_or_else(|| "md n/a".to_string());
        let hooks = agent
            .hooks
            .as_ref()
            .map(|item| format!("hooks {:?}", item.state))
            .unwrap_or_else(|| "hooks n/a".to_string());
        println!(
            "{:<28} {:<10} {} | {} | {}",
            agent.id,
            status_label(agent.status),
            skills,
            instructions,
            hooks
        );
    }
    if !report.problems.is_empty() {
        println!();
        for problem in &report.problems {
            println!("- {}", problem.title);
        }
    }
}

fn print_plan(plan: &SyncPlan, enabled: &[PlanGroupKind]) {
    println!(
        "Plan ({}): {} links, {} writes, {} backups, {} unlinks, {} lock drops; backups use .pre-sync-{}.bak",
        plan.scope,
        plan.summary.links,
        plan.summary.writes,
        plan.summary.backups,
        plan.summary.unlinks,
        plan.summary.drops,
        plan.stamp
    );
    for group in &plan.groups {
        let on = enabled.contains(&group.kind);
        println!();
        println!(
            "[{}] {} ({} change{})",
            if on { "x" } else { " " },
            group.title,
            group.change_count,
            if group.change_count == 1 { "" } else { "s" }
        );
        for op in &group.ops {
            if op.verb == PlanVerb::Keep {
                continue;
            }
            let target = match (&op.target, op.verb) {
                (Some(target), PlanVerb::Link) => format!(" -> {target}"),
                (Some(target), PlanVerb::Backup) => format!(" -> {target}"),
                (Some(target), PlanVerb::Drop) => format!(" {target}"),
                _ => String::new(),
            };
            let note = op
                .note
                .as_ref()
                .map(|note| format!("  ({note})"))
                .unwrap_or_default();
            println!("  {:<7} {}{}{}", verb_label(op.verb), op.path, target, note);
        }
    }
}
